File: packages/python/tests/integration/binding/on_file_rename_test.py
"""Binding-tier red tests for #570: the per-file row seam is `on_file`, not `extract`.

One name for the seam on every surface: `on-file` in TOML, `on_file` in the
Python SDK. The old `extract` keyword is a hard break (no deprecation alias).
"""

import json
import os

import pytest

from dirsql import DirSQL, Table


def _rows(path):
    return [
        {
            "id": os.path.basename(os.path.dirname(path)),
            "body": row["body"],
            "author": row["author"],
        }
        for line in open(path, encoding="utf-8").read().splitlines()
        for row in [json.loads(line)]
    ]


def describe_table_on_file():
    @pytest.mark.asyncio
    async def it_builds_tables_via_the_on_file_keyword(jsonl_dir):
        db = DirSQL(
            jsonl_dir,
            tables=[
                Table(
                    ddl="CREATE TABLE comments (id TEXT, body TEXT, author TEXT)",
                    glob="comments/**/index.jsonl",
                    on_file=_rows,
                ),
            ],
        )
        await db.ready()
        results = await db.query("SELECT DISTINCT id FROM comments")
        assert {r["id"] for r in results} == {"abc", "def"}

    def it_rejects_the_old_extract_keyword():
        with pytest.raises(TypeError):
            Table(
                ddl="CREATE TABLE t (n INTEGER)",
                glob="**/*.json",
                extract=lambda path: [],
            )

File: packages/ts/tests/integration/binding/on-file-rename.test.ts
// Binding-tier red tests for #570 (real napi core, real fs, nothing mocked):
// the per-file row seam is `onFile`, not `extract`. The old `extract`
// spelling is a hard break (no deprecation alias).
import { readFileSync } from "node:fs";
import { mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

describe("the onFile table seam against the real core", () => {
  let dir: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-test-"));
    await mkdir(join(dir, "data"), { recursive: true });
    await writeFile(
      join(dir, "data", "users.json"),
      JSON.stringify([
        { name: "Alice", age: 30 },
        { name: "Bob", age: 25 },
      ]),
    );
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  it("indexes rows produced by a table's onFile callback", async () => {
    const db = new DirSQL({
      root: dir,
      tables: [
        {
          name: "users",
          ddl: "CREATE TABLE users (name TEXT, age INTEGER)",
          glob: "data/users.json",
          onFile: (filePath: string) =>
            JSON.parse(readFileSync(filePath, "utf8")),
        },
      ],
    });

    const rows = await db.query("SELECT name FROM users ORDER BY name");
    expect(rows.map((r) => r.name)).toEqual(["Alice", "Bob"]);
  });
});
